4단계 - 요금 조회 PR 요청드립니다.

earth-h opened the pull request:
안녕하세요, 영훈님! 4단계 PR 피드백 요청드립니다 :)

이번 과제는.. 체감상 제일 어려웠던 것 같습니다 😭
테스트코드도 열심히 짜보긴 했는데,, 잘 진행했는지 걱정이 되네요..!

이번 4단계에서 중점적으로 생각하고 진행한 부분 간단히 정리하여 공유드립니다! 
1. 경로 조회 시에 비회원인지 여부를 판단하고자 `@AuthenticationPrincipal` 어노테이션에 required 필드를 생성해주었습니다. 해당 필드가 false일 때는 nonMember()라는 정적 팩토리 메소드를 통해 성인 나이를 가진 LoginMember 객체를 반환하였습니다.
2. 나이별 요금 계산과 거리별 요금 계산은 이전에 로또 과제를 진행했을 때처럼 enum을 통해 관리하였습니다.  
- 사실, Distance의 minKm는 Distance 객체로 표현해주고 싶었지만, minKm이 0일 수 있어 validation 걸리는 부분이 생겨 int로.. 진행하였습니다 ㅠ (Distance의 validation을 제거하기에는 구간의 거리는 0이 될 수 없어 이렇게 진행했습니다..!)
- 나이의 경우 아기/어린이/청소년/어른으로 진행하여서, 노인에 대한 요금은 따로 고려하지 않았습니다.
3. 최단 거리를 구한 후 요금을 구하는 순서는 아래와 같이 진행했습니다.
  - 최단 거리 내 역들이 속한 노선들의 노선별 추가 요금 중 가장 큰 값 확인
  - 노선별 최대 추가 요금과 최단 거리를 가지고 Fare(요금) 도메인에 메시지를 보내 요금 확인
  - 만약, 성인이 아니라면 반환받은 Path 내 요금을 연령별 요금 정책에 맞게 변환

------

이번 과정을 진행하면서 여쭤보고자 하는 부분은 다음과 같습니다.
1. 요금을 계산할 때에 subtract(), multiply() 등등을 할 때 매번 새로운 객체를 생성해주었는데요..! VO다 보니, 그렇게 진행했는데, 이렇게 했을 때 너무 Fare 객체가 많아질까 염려가 좀 되었습니다 ㅠ 이럴 때는 어떤식으로 하면 좋을지 조언을 구하고자 합니다.
2. 위 진행사항에서도 공유드린 부분이지만, DistanceFarePolicy에서 minKm 타입을 int로 진행했습니다. 이걸 Distance로 하고 구역 사이의 Distance가 0이 안되어야 하는 부분은 Section에서 vadliation을 따로 거는게 더 나을지..? 여쭤보고 싶습니다!

느끼기에는 저번 과제보다 시간도 많이 투자하고 고민도 많이하였는데, 변경된 내용은 뭔가 더 적은거 같아 의아한 점이 있네요 ㅎㅎ..

인수 테스트 기반 TDD 과제 리뷰 피드백 감사합니다 :)

earth-h committed:
refactor(service): 동일한 로직 메소드 추출(피드백 반영)

earth-h committed:
refactor(domain): 현재 로직 상, Favorite을 레포지토리에서 조회한 후 source/target 지하철역은 조회하나 즐겨찾기의 주인인 회원정보를 조회하지 않기 때문에 FetchType을 LAZY로 설정(피드백 반영)

earth-h committed:
refactor(service): command 형태의 서비스는 @Transactional 설정(피드백 반영)

earth-h committed:
test(service): 빠져있던 테스트 추가(중복 즐겨찾기 예외 테스트)

earth-h committed:
refactor(dto): 접근제어자가 빠져있어 추가함(피드백 반영)

earth-h committed:
refactor(service): 토큰이 유효하지 않을 때 Exception을 throw하여 UNAUTHROIZED httpstatus 반환(피드백 반영)

earth-h committed:
docs(guide): 4단계 - 요금 조회 요구사항 정리

earth-h committed:
feat(domain): 거리 비교 메소드 추가(파라미터로 넘어온 거리보다 크면 true 반환)

earth-h committed:
feat(domain): 나이 비교 메소드 추가(파라미터로 넘어온 나이보다 크거나 같으면 참 반환하는 메소드)

earth-h committed:
feat(domain): 주어진 거리로 나눈 후 소숫점 첫자리에서 절상하여 int 반환하는 메소드 생성

earth-h committed:
feat(domain): 거리의 크기 비교 메소드 변경(큰지 판단하지 않고 크거나 같은지 판단하도록 수정)

earth-h committed:
feat(domain): 거리의 뺄셈 메소드 추가 및 덧셈 메소드 인자 타입 변경(Distance 타입이 아닌 int형으로 받도록 함)

earth-h committed:
feat(domain): 거리 비교 메소드 변경(크거나 같다가 아닌 크다에 대한 메소드로 변경) 및 테스트 보충

earth-h committed:
feat(domain): 거리 및 연령에 따른 요금제 생성 도메인 로직 작성

Fare
- createFare()라는 정적 팩토리 메소드를 통해 먼저 거리에 따른 추가 요금을 계산한 후 연령에 따라 할인율을 적용합니다.
  - 로그인한 사용자가 아닐 경우, 성인 요금으로 사용하면 되기 때문에 따로 연령에 따른 계산을 추가하지 않은 정적 팩토리 메소드 또한 함께 만들었습니다.

AgeFarePolicy
- 연령별 요금 정책 enum 클래스
- 각 연령별로 최소 나이와 성인 요금 기준 내야 할 비율을 함께 넣었습니다.
- 또한, 각 연령별로 비율을 생각하기 전에 공제해야 할 금액이 있어 해당 공제금액 역시 넣어두었습니다.
- 요금 계산은, 먼저 기존 금액에서 공제금액을 뺀 후 비율을 곱하는 식으로 진행했습니다.
  - 소숫점에서 절상하여 나타냈습니다.

DistanceFarePolicy
- 거리별 요금 정책 enum 클래스
- 각 거리별로 최소 거리(해당 거리 초과해야 해당함)를 넣었고, 추가 요금 인터벌과 해당 인터벌마다 얼마씩 추가 부과되는지 넣었습니다.
  - 추가적으로, 거리별로 구간이 A - B - C가 있다고 할 때 C 구간에 속한 거리는 A - B 사이 구간의 추가 요금은 필수적으로 들어가기 때문에 기본적인 추가금액(basicAdditionalFare)을 넣어주었습니다.
- 거리에 따른 추가 요금 계산은, 먼저 거리에서 현 구간 최소 거리를 뺀 후 인터벌로 나누어 소숫점 절상하여 몇 번 추가 요금을 적용할지 체크했습니다.
  - 이렇게 개수가 나오면 인터벌 비용만큼 곱해주고, 기본적인 추가금액을 반환하였습니다.

earth-h committed:
feat(domain): 노선 내 추가 요금 필드 생성 및 필드 생성으로 인한 test fixture 수정

earth-h committed:
feat(domain): 거리를 구할 때에는 section의 upStation/downStation의 순서를 보지 않기 때문에 구간 내 역이 동일한지 판단하고자 isContainStationsInAnyOrder() 메소드 생성

earth-h committed:
feat(domain): Path 내 요금 필드 추가하여 PathResponse에 요금 전달할 수 있도록 로직 수정 및 최단거리 조회 시, 요금 역시 함께 조회하도록 서비스 수정

earth-h committed:
fix(dto): DTO 내 getter가 존재하지 않아 인수테스트 깨지는 부분 수정 반영

earth-h committed:
feat(dto, test): 노선 생성 요청 시, 노선별 추가 금액을 넣을 수 있도록 dto 수정 및 최단 경로 조회 시 해당 금액 반환 확인 테스트 추가

earth-h committed:
refactor(dto): dto 내 한 줄에 점 하나만 찍도록 한줄 띄어씀

earth-h committed:
refactor(domain): LoginMember 클래스에 대해 Age만 등록할 수 있는 멤버가 아닌 객체 생성하고자 정적 팩토리 메소드 생성

earth-h committed:
feat(domain, service, controller): Path 조회 시 로그인 여부에 따라 나이 체크하여 요금 반환하도록 함

AuthenticationPrincipal
- required 필드 생성하여, 기본적으로 true를 갖고 false를 설정할 수 있도록 함

AuthenticationPrincipalArgumentResolver
- required 필드값에 따라 LoginMember 객체 생성 방식 분기처리
  - nonMember()의 경우 Age를 성인 나이로 등록한 객체로 로그인하지 않은 사용자일 때 사용함

PathService
- findShortestPath의 파라미터 변경하여 연령별 요금 정책을 적용할 수 있도록 함
- Path에 속한 요금을 연령에 따라 convert할 수 있도록 로직 구성

Path
- Path에 존재하는 요금을 연령에 따라 계산할 수 있도록 메소드 생성

PathController
- 토큰을 통해 로그인유저인지 식별할 수 있도록 @AuthenticationPrincipal 설정

PathService
- 최단경로 찾는 메소드의 인자에 연령별 요금 정책을 추가하여 성인이 아닐 때에는 추가적으로 요금 계산하도록 함

earth-h committed:
docs(guide): README.md 내 진행 완료한 요구사항 체크

earth-h committed:
refactor(controller): 생성자 주입 대상 객체에 final 설정

earth-h committed:
refactor(domain): nonMember는 비회원 전용이므로, 여러 번 생성되지 않게 static final로 생성함

chae-yh commented on README.md:
````diff
@@ -429,3 +429,51 @@ fun RequestSpecification.즐겨찾기_생성_요청(
 ```
+
+## 4단계 - 요금 조회
+### 요구 사항
````
기능 요구 사항 작성 👍

chae-yh commented on src/main/java/nextstep/subway/auth/domain/LoginMember.java:
```diff
@@ -1,20 +1,32 @@
 public class LoginMember {
+
+    private static final Age ADULT = AgeFarePolicy.ADULT.getMinAge();
+    private static final LoginMember nonMember = new LoginMember(ADULT);
```
nonMemer()에서만 사용하는 것 같은데, 이 부분을 필드로 가지고 있는 이유를 여쭤봐도 될까요..?

earth-h replied:
nonMember()에서만 사용하는데, nonMember() 메소드 호출시마다 새로운 new LoginMember(ADULT)를 호출하지 않고자 하나로 선언하여 사용하였습니다 :) 

혹, 이렇게 하는거 보다, 매번 정적 팩토리 메소드 nonMember() 호출 시 새로운 어른 유저를 만드는게 나은 방식이라면 알려주시면 감사하겠습니다 :)

chae-yh replied:
메모리를 적게 사용하는 방향으로 작성을 해주시는 것 같네요 👍
저도 이 부분은 크게 이슈 없을 것 같습니다..!

chae-yh commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -26,6 +27,11 @@ public boolean supportsParameter(MethodParameter parameter) {
+        AuthenticationPrincipal parameterAnnotation = parameter.getParameterAnnotation(AuthenticationPrincipal.class);
+        boolean credentialRequired = parameterAnnotation.required();
+        if(!credentialRequired && credentials == null) {
+            return LoginMember.nonMember();
```
비회원 처리 👍

chae-yh commented on src/main/java/nextstep/subway/fare/domain/DistanceFarePolicy.java:
```diff
@@ -0,0 +1,36 @@
+    ZERO_KM(0, Distance.from(10), Fare.from(0), Fare.from(0))
+    ;
+
+    private int minKm;
```
질문 주신 내용 여기서 답변 드립니다...!

`위 진행사항에서도 공유드린 부분이지만, DistanceFarePolicy에서 minKm 타입을 int로 진행했습니다. 이걸 Distance로 하고 구역 사이의 Distance가 0이 안되어야 하는 부분은 Section에서 vadliation을 따로 거는게 더 나을지..? 여쭤보고 싶습니다!`

이 부분 관련해서는, 1) 새로운 요구사항을 만족시키기 위해 2) Section의 사이는 0이 될 수 없는 validation이 Section에 있는 것은 자연스러우므로, Distance에서는 크기 비교 메서드만 제공하고, Section에서 체크해줄 수 있을 것 같습니다.
아니면 현재 Fare 와 Line 도메인을 분리하였으므로, Fare에서 새로운 Distance를 선언해서 사용해도 될 것 같아요...! 같은 '거리'이지만 도메인에 따라 다른 의미를 가질 수 있으므로, 따로따로 가지고 있어도 좋을 것 같습니다..!
추가로, 관련해서는 뒤의 미션의 DDD를 학습하시면서 비슷한 상황 및 개념을 배우시게 될 것 같습니다..! (비슷한 개념이지만 다른 도메인에서는 다른 객체로 사용 등)

earth-h replied:
답변 감사합니다 !

우선, DDD 학습 관련 다른 객체로 사용하는 부분은 아직 배우지 않은 상태이니, validation을 변경하는 방식으로 DistanceFarePolicy에서 Distance 객체를 사용할 수 있도록 수정해보겠습니다 :)

chae-yh commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -34,10 +34,29 @@ public Distance subtract(Distance distance) {
+        if(isZero(distance)) {
+            throw new IllegalArgumentException(ErrorCode.나누는_값은_0일_수_없음.getErrorMessage());
+        }
+        return (int) Math.ceil((double) this.distance / distance.distance);
```
부동 소수점 계산은 오차가 발생할 수 있습니다. BigDecimal을 써보는 것은 어떨까요..?

earth-h replied:
아하 넵넵 그렇겠네요...! 부동 소수점 이슈가 있을 수 있죵..! 

지하철 실제 요금표 보면서, 사실상 소수점이 없구나 하는 생각이 들어서 int로 진행했던 부분이라 
요거는 말씀해주신대로 BigDecimal로 수정해보겠습니다 :) 

피드백 감사합니다 !

chae-yh replied:
계산 결과를 BigDecimal로 소수점 처리해주신 것 같습니다..! 다만, 부동소수점 연산(여기서는 나눗셈 연산)에서 오차가 발생할 것 같아요..! 
아예 계산 시점부터 BigDecimal을 사용해줘도 좋을 것 같습니다 ㅎㅎ

earth-h replied:
헉 요금만 bigdecimal로 계산을 수정하였었네요..!
이미 머지가 되어서 요건은 제 레포에만 수정해서 반영해두겠습니다 :)
한 번 더 체크해주셔서 감사합니다 ! 

```java
    public int divideAndCeil(Distance distance) {
        if(isZero(distance)) {
            throw new IllegalArgumentException(ErrorCode.나누는_값은_0일_수_없음.getErrorMessage());
        }
        return BigDecimal.valueOf(this.distance)
                .divide(BigDecimal.valueOf(distance.distance), 0, RoundingMode.CEILING).intValue();
    }

```

chae-yh commented on src/main/java/nextstep/subway/fare/domain/Fare.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public Fare multiply(int count) {
+        return new Fare(fare * count);
```
질문 주신 내용 여기서 답변 드립니다..!

`요금을 계산할 때에 subtract(), multiply() 등등을 할 때 매번 새로운 객체를 생성해주었는데요..! VO다 보니, 그렇게 진행했는데, 이렇게 했을 때 너무 Fare 객체가 많아질까 염려가 좀 되었습니다 ㅠ 이럴 때는 어떤식으로 하면 좋을지 조언을 구하고자 합니다.`

사실 이 부분의 의도를 정확히 이해를 하지는 못했습니다..ㅠ 객체가 많아져서 메모리 사용량이 늘어나는 점에 대한 염려였을까요...?

earth-h replied:
넵넵 메모리 사용량에 대한 염려였습니다..!

혹시 메모리 사용량은 여기서는 고려하지 않아도 되는 포인트였을까요?? 최근에 OOM 관련해서 생각을 자꾸 하다보니.. new만 보면 메모리가 걱정되더라구요😭

GC에 의해 안 쓰는 객체는 사라질거같긴한데, 위와 같이 진행하는데에 문제점이 없을지 궁금하였습니다 !

chae-yh replied:
우선 클린코드의 목적상 메모리 사용량 / 성능 등에 대한 고려는 우선순위를 뒤로 두셔도 될 것 같아요. 참고로, 직관적인 설계와 성능은 종종 트레이드 오프 관계이고, 마틴 파울러는 리팩토링에서 성능이 차라리 성능이 떨어지더라도 직관적으로 작성하라고 말하고 있습니다.

그리고, OOM에 대한 고려를 계속 해주시면서 작업해주시는 것은 좋네요 👍 다만, 대체적으로 봤을 때 계산 결과로 만들어진 새로운 객체를 사용하게 될 것이므로, 기존 객체는 GC에 의해 사라질 것으로 생각됩니다..! 그리고, 개인적으로는 이런 부분은 배포 후에 heap 메모리에 대한 모니터링으로도 방어가 가능한 범위라고 생각이 들어서 해당 진행방향도 크게 이슈가 될 것 같지는 않습니다..!

earth-h replied:
답변 감사합니다 ! 

중점을 어디에 두고 하느냐에 따라 바뀌는 부분들인데, 제가 서로 트레이드 오프 관계인 부분에 대해 함께 고민을 하여서 오히려 더 생각이 많아졌던거 같네요..! 업무 상황에서도 이 부분 머리속에 넣어두고, 선택과 집중을 한 후 우선순위가 밀린 부분에 대해 어떻게 대처할 지 등을 고민해봐야겠습니다 :)

chae-yh commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -25,13 +27,20 @@ public PathService(LineRepository lineRepository, StationRepository stationRepos
-    public PathResponse findShortestPath(Long source, Long target) {
+    public PathResponse findShortestPath(AgeFarePolicy ageFarePolicy, Long source, Long target) {
+        Path shortestPath = findShortestPathInGraph(source, target);
+        if(ageFarePolicy != AgeFarePolicy.ADULT) {
```
요금과 나이에 관련된 정책들은 AgeFarePolicy 등에 들어가면 어떨까요...? 관련한 정책이 수정되었을 때 이 부분이 누락될 수 있을 것 같습니다..!

earth-h replied:
AgeFarePolicy에 넣으려다가,
`shortestPath.convertFareByAgeFarePolicy(ageFarePolicy);`
요 부분이 이미 존재하여서, Path에서 ageFarePolicy를 참조하는 부분이 있는데, 
AgeFarePolicy에서 Path를 참조하면 순환 참조가 될 거 같다는 생각에 넣지 않았던 부분인데요!

말씀해주신 `관련한 정책이 수정되었을 때 이 부분이 누락될 수 있을 것 같다`라는 부분에는 동의하여서,
Path 도메인 로직 중 convertFareByAgeFarePolicy() 메소드 내에서 판단하여서 요금을 수정해보는 방향으로 수정해보도록 하겠습니다!

chae-yh commented on src/main/java/nextstep/subway/path/domain/DijkstraPathFinder.java:
```diff
@@ -14,6 +20,7 @@ public class DijkstraPathFinder implements PathFinder {
 
     private final WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph<>(
             DefaultWeightedEdge.class);
+    private final List<Section> sections = new ArrayList<>();
```
요금 계산을 위해 추가하신 것 같습니다. 그런데 이렇게 별개의 필드로 있기보다는, 바로 위의 WeightedMultigraph의 DefaultWeightedEdge를 변경해보는 것은 어떨까요...?

earth-h replied:
이 피드백 관련해서 댓글을 남겼다고 생각했는데 없었네요..!

DefaultWeightedEdge를 고칠 수 없는 값인 줄 알고, Sections를 추가했던 부분인데요!
영훈님 피드백이 힌트가 되어, 수정해볼 수 있었습니다 ㅎㅎ 감사합니다 :) 

최초에 작성할 때, 미금역-정자역과 같이 동일한 up/downStation 구역이 서로 다른 노선에 존재할 수 있는데, 
이때 거리가 좀 다를 수 있다고 판단해서 이 부분은 어떻게 찾을 수 있을까 고민하였었는데요.!
그 부분을 확인할 방법을 찾지 못하였는데, DefaultWeightedEdge 대신에 Section을 쓰게 되어 그 부분도 해결할 수 있었습니다 !

이 피드백 덕분에 jgrapht에 대해 좀 더 알아볼 시간이 되었던거 같아서 너무 좋았습니다 ! 🤩

chae-yh commented on src/test/java/nextstep/subway/fare/domain/AgeFarePolicyTest.java:
```diff
@@ -0,0 +1,76 @@
+import org.junit.jupiter.params.provider.ValueSource;
+
+@DisplayName("연령별 요금 정책 도메인 테스트")
+class AgeFarePolicyTest {
```
요금 정책에 대한 단위 테스트 👍

chae-yh commented on src/test/java/nextstep/subway/favorite/application/FavoriteServiceTest.java:
```diff
@@ -131,6 +131,22 @@ void createFavoriteThrowErrorWhenTargetStationIsNotExists() {
                 .hasMessage(ErrorCode.존재하지_않는_역.getErrorMessage());
     }
 
+    @DisplayName("즐겨찾기 생성 시, 이미 존재하는 즐겨찾기라면 예외를 발생시킨다.")
```
피드백 반영 👍

chae-yh commented on src/test/java/nextstep/subway/path/application/PathServiceTest.java:
```diff
@@ -98,6 +102,131 @@ void findShortestPath() {
 
+    @DisplayName("최단 경로에 속한 노선에 추가 요금이 존재하면, 해당 금액이 추가된 요금이 반환된다.")
+    @Test
+    void findShortestPathWithLineFare() {
```
요금 계산에 대한 mock 테스트 👍

chae-yh commented on src/main/java/nextstep/subway/fare/domain/AgeFarePolicy.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Arrays;
+import nextstep.subway.member.domain.Age;
+
+public enum AgeFarePolicy {
```
이 부분은 의견여쭤보고 싶습니다. Fare라는 도메인이 추가가 되었는데, 이 부분은 노선에 대한 요금이기 때문에 line 밑의 domain으로 가는 것은 어떤가요...?
Fare는 Service Layer / Controller Layer를 갖지 않는데 이러한 구성을 갖게 된 사유도 궁금합니다..!

earth-h replied:
Fare가 노선당 추가요금일수도 있고, 구역간 거리에 따른 요금일 수 있어서 였는데요!
생각해보니 노션과 구역은 line 패키지였죠.. 이 부분은 생각하다가 좀 꼬였던거 같네용..!

line쪽으로 옮겨보겠습니다 ! 피드백 감사합니다~

chae-yh reviewed (changes requested):
지수님 안녕하세요. 피드백이 다소 늦어지게 되어 죄송합니다.
그리고 마지막도 코드 잘 작성해주셨습니다 👍 이대로 바로 merge를 해도 좋겠지만 조금만 더 고민해주시면 좋을만한 내용들이 있고, 질문 주신 내용에 대한 답변과 문의 내용도 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

earth-h committed:
refactor(domain): Fare, AgeFarePolicy, DistanceFarePolicy는 노선 및 구간에 대한 요금 도메인이므로, 패키지를 line으로 리팩토링(피드백 반영)

earth-h committed:
refactor(domain): Distance는 거리이므로, 0보다 작을때만 validation 설정하고, Path나 Section과 같이 Distance를 사용해 만들어진 거리에 대해서 0일 때 validation 걸리도록 추가 validation 설정함(피드백 반영)

Distance validation 변경 사유
- DistanceFarePolicy 내에서 Distance 객체를 사용하고자 하는데, 0km 이상일 때에 대해 체크해야함
- 그런데 기존에 Distance는 0보다 커야하는 validation이 있어 제대로 사용이 불가하여, Distance라는 이름에 맞게 음수만 막고, Distance를 쓰는 영역에서 추가 validation 넣음

earth-h committed:
refactor(service, domain): PathService 내에서 연령별 요금 계산을 위해 조건문을 건 부분을 도메인 로직을 뺌(피드백 반영)

추후 정책이 변경되었을 때 누락되지 않고 변경되도록 하기 위함

earth-h committed:
refactor(domain): DijkstraPathFinder 내 존재하던 sections 리스트 대신, graph 내 edge로 Section을 사용하도록 함

DijkstraPathFinder
- DefaultWeightedEdge를 edge로 사용하는 대신에 Section을 edge로 사용함
- edge가 Section으로 변경됨에 따라, graph 내에서 특정 upStation, downStation으로 이루어진 section 리스트들을 찾아 그 중에 가장 짧은 거리를 갖는 구역을 반환할 수 있도록 메소드 리팩토링

Section
- DefaultWeightedEdge를 extends 해야 AbstractBaseGraph에서 assert (e instanceof DefaultWeightedEdge) 부분에 걸리지 않아 해당 부분 추가함
- 인자로 주어진 구역들 중, 구역의 거리가 가장 짧은 구역을 찾는 메소드 추가
  - PathFinder에서 사용

earth-h committed:
refactor(domain): 부동소수점 계산 시 오차가 발생할 수 있어 BigDecimal로 Fare 필드 타입 변경(피드백 반영)

Fare
- Fare 필드에 대해 BigDecimal로 변경
- BigDecimal 사용에 따른 사칙연산 로직 수정

LineResponse & PathResponse
- 응답 DTO에서 fare 반환 시, BigDecimal로 반환하므로, 소수점 보이지 않도록 RoundingMode 추가

FareTest & PathAcceptanceTest & PathTest
- Fare 필드 타입 변경에 따른 테스트 로직 수정

earth-h commented:
안녕하세요! 주신 피드백들 바탕으로 수정해보았습니다 :) 

- nonMember() 관련 필드 가지고 있는 부분은 사유를 댓글로 남겨두었습니다.
- Fare 패키지 위치 변경 -> Line
- Distance validation 처리 방식 변경(여쭤봤던 부분 답변 받은 것 참고)
- 부동 소수점 계산 오차 발생 가능에 따른 요금 필드 BigDecimal 변경
- PathService 내에서 연령에 따른 요금 변경 로직 domain으로 이동
- jgrapht 사용 시, 요금 계산을 위해 추가된 sections 필드 대신 WeightedMultigraph의 edge로 Section 등록

좋은 피드백 주셔서 감사합니다 !

chae-yh reviewed (approved):
지수님 안녕하세요! 피드백 반영 잘 해주셨습니다 💯 질문 주신 내용 등에 대해서 의견을 코멘트로 남겨드렸고, 그 외에는 더 드릴 피드백이 없을 것 같습니다 ㅎㅎ

여기까지가 제가 드리는 마지막 피드백일 것 같아요..! step1부터 코드 깔끔하게 잘 작성해주셔서, 코드를 읽는 것도 꽤 즐거웠고 개인적으로도 많이 배우는 좋은 기회였던 것 같습니다. 반대로 지수님께도 제가드린 피드백이 조금이라도 도움이 되었다면 좋겠습니다 : ) 그리고, 남은 미션들도 지금처럼 꼼꼼히 열정적으로 진행해주신다면 더 많은 성장을 할 수 있을거라 믿어 의심치 않습니다..! 

그 동안 고생 많으셨습니다. 감사합니다. : )
